Tests/ApolloTests/Cache/WatchQueryTests.swift

```diff
@@ -669,7 +669,7 @@ class WatchQueryTests: XCTestCase, CacheDependentTesting {
       wait(for: [serverRequestExpectation, initialWatcherResultExpectation], timeout: Self.defaultWaitTimeout)
     }
     
-    let numberOfFetches = 1000
+    let numberOfFetches = 10
```

Comment:
Do we have reasonable confidence that only 10 concurrent fetches vs. say 100 (I get that 1000 causes timeouts) is enough to sufficiently push watchers?

Reply:
Yeah i think the issue is the CI machines are heavily memory-constrained (sometimes more so than actual iOS devices!), so that may be part of what's causing the slowness.

Reply:
> This test is configured to use InMemoryNormalizedCache too which I'd expect to be quick.

The `SQLiteWatchQueryTests` actually subclass these tests and provide the SQLite cache. So these tests are run both against the InMemory and SQLite caches.

> Do we have reasonable confidence that only 10 concurrent fetches vs. say 100 (I get that 1000 causes timeouts) is enough to sufficiently push watchers?

There is no known race condition encountered in the past that this test is trying to prevent regressing. To me, this test should be testing behavior of the watcher completion block when multiple fetches hit the cache. This should be testing that reproducible behavior. Trying to get a unit test to check for a race condition when there was no known race condition to test seems futile. These could be happening concurrently or serially. Either way, they test that behavior.

The way you test a race condition is different. If some sort of race condition ever shows up, there is no way to have any confidence this test will catch it with 100, 1000, or 10,000 fetches. In the case of a known race condition, we should write a reproducing test; use it to figure out the source of the race condition; come up with a fix; then write a second test that tests the actual behavior written to prevent the race condition; put comments on both of those tests to indicate the race condition that they check for and link it to an issue so we know what to refer back to when we edit it in the future.

I am open to having my opinion changed, but I'm currently philosophically opposed to having "shotgun" tests that just do a bunch of things in the hopes that they may catch a race condition or a bug if they exist -- at least in the unit test target.

If we want to achieve that goal, I think the proper way to do that is to have stress tests that put the system under large load. These should:
  - Be integration tests, **not unit tests**.
  - Check for increased memory load (checking for memory leaks)
  - Check for increased time over iterations (checking for data/object explosion)
  - Be ran periodically on a cron job **not with unit tests**

These shouldn't be run with unit tests because they will inherently be long running, integration style tests, with failures that will likely be transient. That is not conducive to the development process. I run unit tests locally _constantly_ as I am developing (Test Driven Development requires it). Unit tests also hold up PRs from being merged. So slow tests, especially ones with transient failures, cause a huge problem for the development flow. 

I am not against testing the things that you both are concerned about here, but we should be writing a much more robust set of integration/performance tests. The unit test suite is not the appropriate place for these in my opinion.

Reply:
If you would like me to try re-writing this test to use a serial queue and wait upon each individual fetch so we can make it more clear that this is not a stress tests, I'll do that. 

I was just trying to quickly get our transient test failures to stop. And doing this concurrently ensures this tests still tests the behavior, while preventing of from being a stress test that overloads our CI machines.

Reply:
> I am open to having my opinion changed, but I'm currently philosophically opposed to having "shotgun" tests that just do a bunch of things in the hopes that they may catch a race condition or a bug if they exist -- at least in the unit test target.

You don't have to convince me of this, I agree with the philosophy.

> If you would like me to try re-writing this test to use a serial queue and wait upon each individual fetch so we can make it more clear that this is not a stress tests, I'll do that.
> 
> I was just trying to quickly get our transient test failures to stop. And doing this concurrently ensures this tests still tests the behavior, while preventing of from being a stress test that overloads our CI machines.

I don't think it's necessary, this change aims for stability and it can do that in keeping with the existing test.

Reply:
Okay. I was just trying to be clear that I wasn't just trying to prevent timeouts. I think that having this test as a stress test of any sort is not the right direction. Especially since we removed a bunch of the complex async locking mechanisms a while ago.

Reply:
I think that's worth filing an issue on to come back to - this is definitely something where there's a lot of places in the tests that are checking for things that were mostly caused by our Promises implementation that got ripped out. I think if you don't think we should be doing this as stress tests, we should definitely rip it out. 

However I'm fine with this as an intermediate step as long as we have some kind of actual intention to come back to it.